code/game/objects/structures/detectiveboard.dm

```diff
@@ -0,0 +1,254 @@
+
+/// Attaching evidences: photo and papers
+
+/obj/structure/detectiveboard/attackby(obj/item/O, mob/user, params)
+	if(!cases.len)
+		to_chat(user, "There are no cases!")
+		return
+	if(istype(O, /obj/item/paper) || istype(O, /obj/item/photo))
+		if(attaching_evidence)
+			to_chat(user, "You already attaching evidence!")
+			return
+		if(!allowed(user))
+			to_chat(user, span_warning("You are not authorized to add notices!"))
+			return
+		attaching_evidence = TRUE
+		var/name = tgui_input_text(usr, "Please enter the evidence name", "Detective's Board")
+		if(!name || name == "")
+			attaching_evidence = FALSE
+			return
+		var/desc = tgui_input_text(usr, "Please enter the evidence description", "Detective's Board")
+		if(!desc || desc == "")
+			attaching_evidence = FALSE
+			return
+
+		if(!user.transferItemToLoc(O, src))
+			attaching_evidence = FALSE
+			return
+		cases[current_case].notices++
+		var/datum/evidence/evidence = new (name, desc, O)
+		cases[current_case].evidences += evidence
+		to_chat(user, span_notice("You pin the [O] to the detective board."))
+		attaching_evidence = FALSE
+		update_appearance(UPDATE_ICON)
+	else
+		return ..()
+
+/obj/structure/detectiveboard/ui_state(mob/user)
+	return GLOB.physical_state
+
+/obj/structure/detectiveboard/ui_interact(mob/user, datum/tgui/ui)
+	ui = SStgui.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "DetectiveBoard", name)
+		ui.open()
+
+/obj/structure/detectiveboard/ui_data(mob/user)
+	var/list/data = list()
+	var/list/data_cases = list()
+	for(var/datum/case/case in cases)
+		var/list/data_case = list("ref"=REF(case),"name" = case.name, "color" = case.color)
+		var/list/data_evidences = list()
+		for(var/datum/evidence/evidence in case.evidences)
+			var/list/data_evidence = list("ref" = REF(evidence), "name" = evidence.name, "type" = evidence.evidence_type, "description" = evidence.description, "x"=evidence.x, "y"=evidence.y)
+			var/list/data_connections = list()
+			for(var/datum/evidence/connection in evidence.connections)
+				data_connections += list(list("ref" = REF(connection))) // TODO: create array of strings
+			data_evidence["connections"] = data_connections
+			switch(evidence.evidence_type)
+				if("photo")
+					var/obj/item/photo/photo = evidence.item
+					var/tmp_picture_name = "evidence_photo[REF(photo)].png"
+					user << browse_rsc(photo.picture.picture_image, tmp_picture_name)
+					data_evidence["photo_url"] = tmp_picture_name
+				if("paper")
+					var/obj/item/paper/paper = evidence.item
+					data_evidence["text"] = ""
+					if(paper.raw_text_inputs && paper.raw_text_inputs.len)
+						data_evidence["text"] = paper.raw_text_inputs[1].raw_text
+			data_evidences += list(data_evidence)
+		data_case["evidences"] = data_evidences
+		data_cases += list(data_case)
+	data["cases"] = data_cases
+	data["current_case"] = current_case
+	data_test = data
+	return data
+
+/obj/structure/detectiveboard/ui_act(action, params)
```

Comment:
There's 2 args missing from here, one of which is the tgui window, where you can get the user of the ui with ``ui.user``, which would be preferable to use over ``usr`` further down.